### ISSUE TITLE
Many deoptimizations/invalidations when a benchmark ends

### ISSUE THREAD

**Issue by eregon** [Shopify/yjit-bench#210]
I've noticed this on TruffleRuby:
```
[engine] opt inval. id=1137  Integer#==                                                                                                                                                         |Timestamp 15414910326666|Src (core):1|Reason hierarchy is unmodified Array
```
And many many other invalidations after the last iteration was measured.

So the good news is it shouldn't affect the captured times.
The bad news is this may cause the process to take a while longer to terminate, may cause more memory to be used due to extra invalidations & compilations, etc.

Given the above means a module was included or prepended to `Integer`, this is almost certainly the side effect of `require 'json'` so late: https://github.com/Shopify/yjit-bench/blob/5b1ec56ad5ff9ba4a159bf91d2711db36cad1d71/harness/harness-common.rb#L60

Maybe we could just serialize to json ourselves to avoid this, given the data we dump is so simple (RUBY_DESCRIPTION and floats). Any thoughts on that?

Requiring json early doesn't seem good, especially for smaller benchmarks.

Of course it's not great the `json` gem does such things (https://github.com/flori/json/blob/248bc5bf7ea597d7a20396a8def855a1988bf490/lib/json/common.rb#LL69C11-L69C18 and https://github.com/flori/json/blob/248bc5bf7ea597d7a20396a8def855a1988bf490/lib/json/pure/generator.rb#L285) but I don't think we can do anything about that here. In fact if it would not `include` but just define the methods directly it wouldn't be an issue for per class & method name caches like in TruffleRuby (and I think in CRuby too).

**Comment by maximecb** [comment]
> Maybe we could just serialize to json ourselves to avoid this

This definitely looks more like a bug in TruffleRuby than a problem with the benchmarking harness.

If there are simple changes that could be made to the json gem to make it more amenable for optimization, I would say take it up with them.

**Comment by eregon** [comment]
> This definitely looks more like a bug in TruffleRuby than a problem with the benchmarking harness.

I would think this also causes invalidations on many Ruby implementations (or those which use inline caches and a JIT at least).
Is there a flag to tell YJIT to show code thrown away due to invalidations like monkey-patching?

For example one must invalidate the inline cache of `someInteger.is_a?(someModule)` if the hierarchy of `Integer` changes.

**Comment by eregon** [comment]
> Does TR take a noticeable amount of time to terminate because of this?

Not really, no, but in this case I was tracing compilation (`--engine.TraceCompilation`) and it's basically causing a ton of noise. Similarly if I would gather compilation statistics I would then see a big number of deops/invalidations which are actually not caused by the benchmark, but by the harness.
I could use yet another harness, but OTOH I would want the same information, so not really a solution.

My initial concern was that it may skew the memory measurement, but  the RSS is collected actually just before `require "json"` so that's fine if that's the main place we measure memory (and would also work if we switch to getrusage()'s maxrss).

In general I think it's valuable that the benchmark harness adds as little work as possible.
Hence I think here it would make a lot of sense to just serialize to JSON ourselves, it should be trivial, I'll try it.

> Either way, the json gem is obviously pretty common, and people do tend to require gems at the top of the file.

Yes, and if it's required early it's no problem for invalidations of course. So here the issue is the harness tries to avoid to load things unrelated to the benchmark (good), but then has the side effect to cause tons of invalidations due to the late require monkey patching/including modules in about every core class. We can easily have the best of both, I'll make a PR.

**Comment by eregon** [comment]
I've looked at usages of the hierarchy assumption, and it's used in only in https://github.com/oracle/truffleruby/blob/master/src/main/java/org/truffleruby/language/objects/IsANode.java.

It is needed for the example I gave above:

> For example one must invalidate the inline cache of `someInteger.is_a?(someModule)` if the hierarchy of `Integer` changes.

But it is not needed for the case `receiver.is_a?(someClass)`, since the superclass chain cannot change.
And that's by far the most common case of course.
The reason it cannot change is non-trivial because Class#initialize can mutate a Class.allocate class's superclass. But `Class.allocate.new` is not allowed hence it's not possible to create an instance of a class where the superclass is not set yet.
This would be a concern for "is A subclass of B" cases though (e.g. Module#<, when the receiver and arguments are classes).

I guess CRuby doesn't do any inline cache inside `is_a?`/`Moduel#<`, hence this is not seen (yet) as a problem on CRuby.

The bottom line is we can indeed fix this specific case in TruffleRuby and that removes most invalidations caused by `require "json"` (477->47), but not as much as when not requiring `json`.
